econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -0,0 +1,1056 @@
+        """
+        Gets the cohort causal effects for each feature as dictionary.
+
+        Dictionary entries for predictions, etc. will be nested lists of shape (d_y, sum(d_t))
+
+        Only for serialization purposes to upload to AzureML
+        """
+        return self._dict_summary(self._cohort_effect_inference(Xtest), props=self._summary_props(alpha),
+                                  kind='cohort', expand_arr=True, drop_sample=True)
+
+    def _local_effect_inference(self, Xtest):
+        assert np.ndim(Xtest) == 2 and np.shape(Xtest)[1] == self._d_x, (
+            "Shape of Xtest must be compatible with shape of X, "
+            f"but got shape {np.shape(Xtest)} instead of (n, {self._d_x})"
+        )
+
+        def inference_from_result(result):
+            est = result.estimator
+            X = result.X_transformer.transform(Xtest)
+            return est.const_marginal_effect_inference(X=X)
+        return inference_from_result
+
+    def local_causal_effect(self, Xtest, alpha=0.05):
+        """
+        Gets the local causal effect for each feature as a pandas DataFrame.
+
+        Parameters
+        ----------
+        Xtest : the samples for which to return the causal effects
+        alpha : the confidence level of the confidence interval
+
+        Returns
+        -------
+        global_effect : pandas Dataframe
+            DataFrame with the following structure:
+
+            :Columns: ['point', 'stderr', 'zstat', 'pvalue', 'ci_lower', 'ci_upper']
+            :Index: ['sample', 'feature', 'feature_value']
+            :Rows: For each feature that is numeric, we have an entry with index
+                   ['{sampleid}', '{feature_name}', 'num'],
+                   where 'num' is literally the string 'num' and feature_name is the input feature name and sampleid is
+                   the index of the sample in Xtest.
+                   For each feature that is categorical, we have an entry with index
+                   ['{sampleid', '{feature_name}', '{cat}v{base}']
+                   where cat is the category value and base is the category used as baseline.
+                   If all features are numerical then the feature_value index is dropped in the dataframe, but not
+                   in the serialized dict.
+        """
+        return self._pandas_summary(self._local_effect_inference(Xtest),
+                                    props=self._point_props(alpha), n=Xtest.shape[0])
+
+    def _local_causal_effect_dict(self, Xtest, alpha=0.05):
+        """
+        Gets the local feature importance as dictionary
+
+        Dictionary entries for predictions, etc. will be nested lists of shape (n_rows, d_y, sum(d_t))
+
+        Only for serialization purposes to upload to AzureML
+        """
+        return self._dict_summary(self._local_effect_inference(Xtest), props=self._point_props(alpha),
+                                  kind='local')
+
+    def _check_feature_index(self, X, feature_index):
```

Comment:
WE might want to rename this because this seems like a safe accessing of the _result for some feature index.